server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -808,6 +808,39 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
 
     }
 
+    ObjectMapper findParentMapper(String leafFieldPath) {
+        ObjectMapper current = this;
+        String pathInCurrent = leafFieldPath;
+
+        while (current != null) {
+            if (current.mappers.containsKey(pathInCurrent)) {
+                return current;
+            }
+
+            // Go one level down if possible
+            var pathComponents = pathInCurrent.split("\\.");
```

Comment:
Nit: do this once before the loop, increment the start element on each iteration?